// coze-coding-dev-sdk@0.5.5/dist/voice/models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TTSConfig = void 0;
exports.buildTTSApiRequest = buildTTSApiRequest;
exports.buildASRApiRequest = buildASRApiRequest;
class TTSConfig {
}
exports.TTSConfig = TTSConfig;
TTSConfig.DEFAULT_SPEAKER = 'zh_female_xiaohe_uranus_bigtts';
TTSConfig.DEFAULT_AUDIO_FORMAT = 'mp3';
TTSConfig.DEFAULT_SAMPLE_RATE = 24000;
TTSConfig.DEFAULT_SPEECH_RATE = 0;
TTSConfig.DEFAULT_LOUDNESS_RATE = 0;
function buildTTSApiRequest(request) {
    const params = {
        user: {
            uid: request.uid,
        },
        req_params: {
            speaker: request.speaker || TTSConfig.DEFAULT_SPEAKER,
            audio_params: {
                format: request.audioFormat || TTSConfig.DEFAULT_AUDIO_FORMAT,
                sample_rate: request.sampleRate || TTSConfig.DEFAULT_SAMPLE_RATE,
                speech_rate: request.speechRate ?? TTSConfig.DEFAULT_SPEECH_RATE,
                loudness_rate: request.loudnessRate ?? TTSConfig.DEFAULT_LOUDNESS_RATE,
            },
        },
    };
    if (request.text) {
        params.req_params.text = request.text;
    }
    else if (request.ssml) {
        params.req_params.ssml = request.ssml;
    }
    return params;
}
function buildASRApiRequest(request) {
    const audioData = {};
    if (request.url) {
        audioData.url = request.url;
    }
    else if (request.base64Data) {
        audioData.data = request.base64Data;
    }
    return {
        user: {
            uid: request.uid,
        },
        audio: audioData,
    };
}
