// coze-coding-dev-sdk@0.5.5/dist/core/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = void 0;
class Config {
    constructor(config) {
        this.apiKey = config?.apiKey || process.env.COZE_WORKLOAD_IDENTITY_API_KEY || '';
        this.baseUrl = config?.baseUrl || process.env.COZE_INTEGRATION_BASE_URL || '';
        this.modelBaseUrl = config?.modelBaseUrl || process.env.COZE_INTEGRATION_MODEL_BASE_URL || '';
        this.retryTimes = config?.retryTimes ?? 3;
        this.retryDelay = config?.retryDelay ?? 1;
        this.timeout = config?.timeout ?? 900000;
    }
    validate() {
        if (!this.apiKey) {
            throw new Error('API key is required. Set COZE_WORKLOAD_IDENTITY_API_KEY or provide apiKey in config.');
        }
    }
    getHeaders() {
        if (!this.apiKey) {
            throw new Error('API key is required. Set COZE_WORKLOAD_IDENTITY_API_KEY or provide apiKey in config.');
        }
        return {
            'Authorization': `Bearer ${this.apiKey}`,
            'Content-Type': 'application/json',
        };
    }
}
exports.Config = Config;
