// coze-coding-dev-sdk@0.5.5/dist/cli/commands/video.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.videoCommand = void 0;
const commander_1 = require("commander");
const index_1 = require("../../index");
const chalk_1 = __importDefault(require("chalk"));
const ora_1 = __importDefault(require("ora"));
const fs = __importStar(require("fs"));
const constants_1 = require("../constants");
const utils_1 = require("../utils");
function parseResolution(size) {
    if (!size) {
        return {};
    }
    const [widthStr, heightStr] = size.split('x');
    const width = parseInt(widthStr);
    const height = parseInt(heightStr);
    if (width === height) {
        return {
            ratio: '1:1',
            resolution: width <= 1024 ? '720p' : '1080p',
        };
    }
    else if (width > height) {
        return {
            ratio: '16:9',
            resolution: width <= 1440 ? '720p' : '1080p',
        };
    }
    else {
        return {
            ratio: '9:16',
            resolution: height <= 1440 ? '720p' : '1080p',
        };
    }
}
exports.videoCommand = new commander_1.Command('video')
    .description('Generate video using AI (synchronous - waits for completion)')
    .option('-p, --prompt <text>', 'Text prompt')
    .option('-i, --image-url <url>', 'Image URL (single or comma-separated two URLs)')
    .option('-s, --size <size>', 'Video resolution (e.g., 1920x1080)')
    .option('-d, --duration <seconds>', 'Video duration (5-10 seconds)', '5')
    .option('-m, --model <model>', 'Model name', 'doubao-seedance-1-0-pro-fast-251015')
    .option('--callback-url <url>', 'Callback URL')
    .option('--return-last-frame', 'Return last frame image', false)
    .option('--watermark', 'Add watermark', false)
    .option('--seed <number>', 'Random seed')
    .option('--camerafixed', 'Fixed camera', false)
    .option('-o, --output <path>', 'Output file path (JSON)')
    .option('--mock', 'Use mock mode (test run)', false)
    .option('-H, --header <header>', 'Custom HTTP header (format: "Key: Value" or "Key=Value", can be used multiple times)', (value, previous) => previous.concat([value]), [])
    .option('--verbose', 'Enable verbose logging to print HTTP request details', false)
    .action(async (options) => {
    const spinner = (0, ora_1.default)('Generating video...').start();
    try {
        const config = new index_1.Config();
        const customHeaders = {};
        const parsedHeaders = (0, utils_1.parseHeaders)(options.header);
        if (parsedHeaders) {
            Object.assign(customHeaders, parsedHeaders);
        }
        if (options.mock) {
            customHeaders[constants_1.RUN_MODE_HEADER] = constants_1.RUN_MODE_TEST;
            console.log(chalk_1.default.yellow('🧪 Mock mode enabled (test run)'));
        }
        const client = new index_1.VideoGenerationClient(config, Object.keys(customHeaders).length > 0 ? customHeaders : undefined, options.verbose);
        const imageUrls = options.imageUrl
            ? options.imageUrl.split(',').map((url) => url.trim())
            : null;
        const { ratio, resolution } = parseResolution(options.size);
        const videoConfig = {
            resolution: (resolution || '720p'),
            ratio: (ratio || '16:9'),
            duration: parseInt(options.duration),
            watermark: options.watermark,
            seed: options.seed ? parseInt(options.seed) : undefined,
            camerafixed: options.camerafixed,
        };
        const contentItems = [];
        if (options.prompt) {
            contentItems.push({
                type: 'text',
                text: options.prompt,
            });
        }
        if (imageUrls) {
            if (imageUrls.length === 1) {
                contentItems.push({
                    type: 'image_url',
                    image_url: {
                        url: imageUrls[0],
                    },
                    role: 'first_frame',
                });
            }
            else if (imageUrls.length === 2) {
                contentItems.push({
                    type: 'image_url',
                    image_url: {
                        url: imageUrls[0],
                    },
                    role: 'first_frame',
                });
                contentItems.push({
                    type: 'image_url',
                    image_url: {
                        url: imageUrls[1],
                    },
                    role: 'last_frame',
                });
            }
            else {
                throw new Error('Only 1 or 2 images are supported');
            }
        }
        if (contentItems.length === 0) {
            throw new Error('Either --prompt or --image-url must be provided');
        }
        const result = await client.videoGeneration(contentItems, {
            model: options.model,
            config: videoConfig,
            callbackUrl: options.callbackUrl,
            returnLastFrame: options.returnLastFrame,
        });
        spinner.succeed(chalk_1.default.green('✓ Video generation completed!'));
        console.log(chalk_1.default.cyan('\n=== Video Generation Result ==='));
        console.log(chalk_1.default.white('Task ID:'), chalk_1.default.green(result.response.id));
        console.log(chalk_1.default.white('Status:'), chalk_1.default.green(result.response.status));
        if (result.videoUrl) {
            console.log(chalk_1.default.white('Video URL:'), chalk_1.default.green(result.videoUrl));
        }
        if (options.returnLastFrame && result.lastFrameUrl) {
            console.log(chalk_1.default.white('Last Frame URL:'), chalk_1.default.green(result.lastFrameUrl));
        }
        if (result.response.error_message) {
            console.log(chalk_1.default.white('Error:'), chalk_1.default.red(result.response.error_message));
        }
        if (result.videoUrl) {
            console.log(chalk_1.default.cyan('\nComplete Video URL:'));
            console.log(chalk_1.default.green(result.videoUrl));
        }
        if (options.returnLastFrame && result.lastFrameUrl) {
            console.log(chalk_1.default.cyan('\nComplete Last Frame URL:'));
            console.log(chalk_1.default.green(result.lastFrameUrl));
        }
        if (options.output) {
            fs.writeFileSync(options.output, JSON.stringify(result.response, null, 2));
            console.log(chalk_1.default.green(`\n✓ Result saved to: ${options.output}`));
        }
    }
    catch (error) {
        spinner.fail(chalk_1.default.red('✗ Video generation failed'));
        console.error(chalk_1.default.red(`Error: ${error.message}`));
        process.exit(1);
    }
});
