// coze-coding-dev-sdk@0.5.5/dist/video/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoGenerationClient = void 0;
const client_1 = require("../core/client");
const exceptions_1 = require("../core/exceptions");
const models_1 = require("./models");
class VideoGenerationClient extends client_1.BaseClient {
    constructor(config, customHeaders, verbose = false) {
        super(config, customHeaders, verbose);
        this.baseUrl = this.config.baseUrl;
    }
    async videoGeneration(contentItems, options) {
        const pollInterval = 5000;
        const maxWaitTime = 300000;
        const model = options?.model || models_1.VideoConfig.DEFAULT_MODEL;
        const requestData = {
            model,
            content: contentItems,
        };
        if (options?.callbackUrl) {
            requestData.callback_url = options.callbackUrl;
        }
        if (options?.returnLastFrame) {
            requestData.return_last_frame = options.returnLastFrame;
        }
        if (options?.config) {
            requestData.config = {
                resolution: options.config.resolution || models_1.VideoConfig.DEFAULT_RESOLUTION,
                ratio: options.config.ratio || models_1.VideoConfig.DEFAULT_RATIO,
                duration: options.config.duration ?? models_1.VideoConfig.DEFAULT_DURATION,
                watermark: options.config.watermark ?? models_1.VideoConfig.DEFAULT_WATERMARK,
                seed: options.config.seed,
                camerafixed: options.config.camerafixed ?? models_1.VideoConfig.DEFAULT_CAMERA_FIXED,
            };
        }
        let taskId = null;
        let retryCount = 0;
        const maxRetries = 3;
        while (retryCount < maxRetries) {
            try {
                const response = await this.request('POST', `${this.baseUrl}/api/v3/contents/generations/tasks`, requestData);
                taskId = response.id;
                if (!taskId) {
                    throw new exceptions_1.APIError('创建视频生成任务失败：响应中缺少任务ID', 500);
                }
                if (this.verbose) {
                    console.log(`视频生成任务创建成功，任务ID: ${taskId}`);
                }
                break;
            }
            catch (error) {
                retryCount++;
                const errorMsg = error instanceof Error ? error.message : String(error);
                if (this.verbose) {
                    console.error(`创建视频生成任务失败（尝试 ${retryCount}/${maxRetries}）: ${errorMsg}`);
                }
                if (retryCount >= maxRetries) {
                    throw new exceptions_1.APIError(`创建视频生成任务失败，已重试${maxRetries}次: ${errorMsg}`, 500);
                }
                if (errorMsg.toLowerCase().includes('rate limit') || errorMsg.includes('429')) {
                    const waitTime = Math.min(2 ** retryCount * 1000, 10000);
                    if (this.verbose) {
                        console.warn(`遇到速率限制，等待 ${waitTime / 1000} 秒后重试...`);
                    }
                    await this.sleep(waitTime);
                }
                else if (errorMsg.toLowerCase().includes('timeout')) {
                    if (this.verbose) {
                        console.warn('请求超时，等待 2 秒后重试...');
                    }
                    await this.sleep(2000);
                }
                else {
                    throw error;
                }
            }
        }
        const startTime = Date.now();
        let pollCount = 0;
        while (Date.now() - startTime < maxWaitTime) {
            pollCount++;
            try {
                const response = await this.request('GET', `${this.baseUrl}/api/v3/contents/generations/tasks/${taskId}`);
                const status = response.status;
                if (this.verbose) {
                    console.log(`任务 ${taskId} 状态检查 #${pollCount}: ${status}`);
                }
                if (status === 'succeeded') {
                    const videoUrl = response.content?.video_url || null;
                    const lastFrameUrl = response.content?.last_frame_url || '';
                    if (!videoUrl) {
                        throw new exceptions_1.APIError(`视频生成成功但响应中缺少视频URL，任务ID: ${taskId}`, 500);
                    }
                    if (this.verbose) {
                        console.log(`视频生成成功，任务ID: ${taskId}, 视频URL: ${videoUrl}`);
                    }
                    return {
                        videoUrl,
                        response,
                        lastFrameUrl,
                    };
                }
                else if (status === 'failed') {
                    const errorMessage = response.error_message || '未知错误';
                    if (this.verbose) {
                        console.error(`视频生成失败，任务ID: ${taskId}, 错误: ${errorMessage}`);
                    }
                    throw new exceptions_1.APIError(`视频生成失败: ${errorMessage}`, 500);
                }
                else if (status === 'cancelled') {
                    if (this.verbose) {
                        console.warn(`视频生成任务被取消，任务ID: ${taskId}`);
                    }
                    return {
                        videoUrl: null,
                        response,
                        lastFrameUrl: '',
                    };
                }
                else if (status === 'queued' || status === 'running') {
                    await this.sleep(pollInterval);
                    continue;
                }
                else {
                    if (this.verbose) {
                        console.warn(`未知的任务状态: ${status}，任务ID: ${taskId}`);
                    }
                    await this.sleep(pollInterval);
                    continue;
                }
            }
            catch (error) {
                const errorMsg = error instanceof Error ? error.message : String(error);
                if (this.verbose) {
                    console.error(`查询任务状态失败，任务ID: ${taskId}, 错误: ${errorMsg}`);
                }
                if (errorMsg.toLowerCase().includes('not found') || errorMsg.includes('404')) {
                    throw new exceptions_1.APIError(`任务不存在或已过期，任务ID: ${taskId}`, 404);
                }
                if (Date.now() - startTime >= maxWaitTime) {
                    throw error;
                }
                await this.sleep(pollInterval);
                continue;
            }
        }
        const elapsedTime = Math.floor((Date.now() - startTime) / 1000);
        if (this.verbose) {
            console.error(`视频生成超时，任务ID: ${taskId}, 已等待: ${elapsedTime}秒`);
        }
        throw new exceptions_1.APIError(`视频生成超时，已等待 ${elapsedTime} 秒，任务ID: ${taskId}`, 408);
    }
    async videoGenerationAsync(contentItems, options) {
        return this.videoGeneration(contentItems, options);
    }
}
exports.VideoGenerationClient = VideoGenerationClient;
