// coze-coding-dev-sdk@0.5.5/dist/image/models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageGenerationResponseHelper = exports.ImageConfig = void 0;
exports.validateImageSize = validateImageSize;
exports.validateImageSizeOrDefault = validateImageSizeOrDefault;
exports.buildImageGenerationApiRequest = buildImageGenerationApiRequest;
class ImageConfig {
}
exports.ImageConfig = ImageConfig;
ImageConfig.DEFAULT_MODEL = 'doubao-seedream-4-5-251128';
ImageConfig.DEFAULT_SIZE = '2K';
ImageConfig.DEFAULT_WATERMARK = true;
ImageConfig.DEFAULT_RESPONSE_FORMAT = 'url';
ImageConfig.DEFAULT_OPTIMIZE_PROMPT_MODE = 'standard';
ImageConfig.DEFAULT_SEQUENTIAL_IMAGE_GENERATION = 'disabled';
ImageConfig.DEFAULT_SEQUENTIAL_IMAGE_GENERATION_MAX_IMAGES = 15;
ImageConfig.MIN_WIDTH = 2560;
ImageConfig.MAX_WIDTH = 4096;
ImageConfig.MIN_HEIGHT = 1440;
ImageConfig.MAX_HEIGHT = 4096;
function validateImageSize(size) {
    if (size === '2K' || size === '4K') {
        return size;
    }
    const match = size.match(/^(\d+)x(\d+)$/);
    if (!match) {
        throw new Error(`Invalid size format. Must be '2K', '4K', or 'WIDTHxHEIGHT' format. Got: ${size}`);
    }
    const width = parseInt(match[1], 10);
    const height = parseInt(match[2], 10);
    if (width < ImageConfig.MIN_WIDTH || width > ImageConfig.MAX_WIDTH) {
        throw new Error(`Width must be between ${ImageConfig.MIN_WIDTH} and ${ImageConfig.MAX_WIDTH}. Got: ${width}`);
    }
    if (height < ImageConfig.MIN_HEIGHT || height > ImageConfig.MAX_HEIGHT) {
        throw new Error(`Height must be between ${ImageConfig.MIN_HEIGHT} and ${ImageConfig.MAX_HEIGHT}. Got: ${height}`);
    }
    return size;
}
function validateImageSizeOrDefault(size) {
    try {
        return validateImageSize(size);
    }
    catch {
        return ImageConfig.DEFAULT_SIZE;
    }
}
class ImageGenerationResponseHelper {
    constructor(response) {
        this.response = response;
    }
    get success() {
        return (!this.response.error &&
            this.response.data.every((item) => !item.error));
    }
    get imageUrls() {
        return this.response.data
            .filter((item) => item.url)
            .map((item) => item.url);
    }
    get imageB64List() {
        return this.response.data
            .filter((item) => item.b64_json)
            .map((item) => item.b64_json);
    }
    get errorMessages() {
        const messages = [];
        if (this.response.error) {
            messages.push(`API Error: ${this.response.error.message || 'Unknown error'}`);
        }
        for (const item of this.response.data) {
            if (item.error) {
                messages.push(`Image generation error: ${item.error.message || 'Unknown error'}`);
            }
        }
        return messages;
    }
}
exports.ImageGenerationResponseHelper = ImageGenerationResponseHelper;
function buildImageGenerationApiRequest(request, model) {
    const validatedSize = validateImageSizeOrDefault(request.size || ImageConfig.DEFAULT_SIZE);
    return {
        model,
        prompt: request.prompt,
        size: validatedSize,
        watermark: request.watermark ?? ImageConfig.DEFAULT_WATERMARK,
        image: request.image,
        response_format: request.responseFormat || ImageConfig.DEFAULT_RESPONSE_FORMAT,
        optimize_prompt_options: {
            mode: request.optimizePromptMode ||
                ImageConfig.DEFAULT_OPTIMIZE_PROMPT_MODE,
        },
        sequential_image_generation: request.sequentialImageGeneration ||
            ImageConfig.DEFAULT_SEQUENTIAL_IMAGE_GENERATION,
        sequential_image_generation_options: {
            max_images: request.sequentialImageGenerationMaxImages ||
                ImageConfig.DEFAULT_SEQUENTIAL_IMAGE_GENERATION_MAX_IMAGES,
        },
    };
}
