// coze-coding-dev-sdk@0.5.5/dist/database/migration.js
"use strict";
/**
 * Database migration module
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateModels = generateModels;
exports.upgrade = upgrade;
const child_process_1 = require("child_process");
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const os = __importStar(require("os"));
const models_1 = require("./models");
const client_1 = require("./client");
/**
 * Fix Drizzle bug: introspect generates numbers beyond JS precision
 * Replace .generatedByDefaultAsIdentity({...}) with .generatedByDefaultAsIdentity()
 */
function fixGeneratedIdentity(filePath) {
    if (!fs.existsSync(filePath))
        return;
    let content = fs.readFileSync(filePath, 'utf-8');
    const fixed = content.replace(/\.generatedByDefaultAsIdentity\(\{[^}]*\}\)/g, '.generatedByDefaultAsIdentity()');
    if (fixed !== content) {
        fs.writeFileSync(filePath, fixed);
    }
}
async function generateModels(options = {}) {
    const { outputPath = models_1.DatabaseConfig.DEFAULT_SCHEMA_PATH, verbose = false, } = options;
    (0, client_1.loadEnv)();
    const dbUrl = (0, client_1.getDbUrl)();
    const outputDir = path.dirname(outputPath);
    if (outputDir && !fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }
    // Use /tmp for drizzle-kit output, config in parent of outputDir for module resolution
    const tempOutDir = fs.mkdtempSync(path.join(os.tmpdir(), 'drizzle-'));
    const configDir = path.dirname(outputDir); // src/storage/database instead of src/storage/database/shared
    const tempConfigPath = path.join(configDir, `drizzle.config.ts`);
    const drizzleConfig = `
import { defineConfig } from 'drizzle-kit';

export default defineConfig({
  dialect: 'postgresql',
  out: '${tempOutDir}',
  dbCredentials: {
    url: '${dbUrl}',
  },
});
`;
    try {
        fs.writeFileSync(tempConfigPath, drizzleConfig);
        if (verbose) {
            console.log('Introspecting database schema...');
        }
        const cmd = `npx drizzle-kit introspect --config=${tempConfigPath}`;
        if (verbose) {
            console.log(`Running: ${cmd}`);
        }
        let output;
        try {
            output = (0, child_process_1.execSync)(cmd, {
                encoding: 'utf-8',
                stdio: 'pipe',
                timeout: 60000,
            });
        }
        catch (err) {
            // execSync throws on non-zero exit code, include stderr in error message
            const stderr = err.stderr?.toString() || '';
            const stdout = err.stdout?.toString() || '';
            throw new Error(`drizzle-kit introspect failed:\n${stderr || stdout || err.message}`);
        }
        // Check for error patterns in output even if exit code is 0
        if (output && (output.includes('error:') || output.includes('Error:'))) {
            throw new Error(`drizzle-kit introspect failed:\n${output}`);
        }
        if (verbose && output) {
            console.log(output);
        }
        // Fix Drizzle bug in generated schema
        const generatedSchema = path.join(tempOutDir, 'schema.ts');
        const generatedRelations = path.join(tempOutDir, 'relations.ts');
        if (fs.existsSync(generatedSchema)) {
            fixGeneratedIdentity(generatedSchema);
            fs.copyFileSync(generatedSchema, outputPath);
            if (verbose) {
                console.log(`✓ Copied schema.ts to ${outputDir}/`);
            }
        }
        if (fs.existsSync(generatedRelations)) {
            const relationsPath = path.join(outputDir, 'relations.ts');
            fs.copyFileSync(generatedRelations, relationsPath);
            if (verbose) {
                console.log(`✓ Copied relations.ts to ${outputDir}/`);
            }
        }
        // Git add generated files
        try {
            (0, child_process_1.execSync)('git add -A', {
                encoding: 'utf-8',
                stdio: 'pipe',
                cwd: process.cwd(),
            });
        }
        catch {
            // Ignore git errors (might not be in a git repo)
        }
        if (verbose) {
            console.log(`✓ Schema generated successfully`);
        }
    }
    finally {
        // Cleanup temp files
        if (fs.existsSync(tempConfigPath)) {
            fs.unlinkSync(tempConfigPath);
        }
        if (fs.existsSync(tempOutDir)) {
            fs.rmSync(tempOutDir, { recursive: true });
        }
    }
}
async function upgrade(options = {}) {
    const { schemaPath = models_1.DatabaseConfig.DEFAULT_SCHEMA_PATH, projectPath = process.cwd(), verbose = false, } = options;
    (0, client_1.loadEnv)();
    const dbUrl = (0, client_1.getDbUrl)();
    const fullSchemaPath = path.isAbsolute(schemaPath)
        ? schemaPath
        : path.join(projectPath, schemaPath);
    if (!fs.existsSync(fullSchemaPath)) {
        throw new Error(`Schema file not found: ${fullSchemaPath}`);
    }
    // Config in parent of schema directory for module resolution
    const schemaDir = path.dirname(fullSchemaPath); // src/storage/database/shared
    const configDir = path.dirname(schemaDir); // src/storage/database
    const tempConfigPath = path.join(configDir, `drizzle.config.ts`);
    const drizzleConfig = `
import { defineConfig } from 'drizzle-kit';

export default defineConfig({
  dialect: 'postgresql',
  schema: '${fullSchemaPath}',
  dbCredentials: {
    url: '${dbUrl}',
  },
});
`;
    try {
        fs.writeFileSync(tempConfigPath, drizzleConfig);
        if (verbose) {
            console.log('Pushing schema changes to database...');
        }
        // Use --force to skip confirmation
        const cmd = `npx drizzle-kit push --force --config=${tempConfigPath}`;
        if (verbose) {
            console.log(`Running: ${cmd}`);
        }
        let output;
        try {
            output = (0, child_process_1.execSync)(cmd, {
                encoding: 'utf-8',
                stdio: 'pipe',
                timeout: 120000,
                cwd: projectPath,
            });
        }
        catch (err) {
            // execSync throws on non-zero exit code, include stderr in error message
            const stderr = err.stderr?.toString() || '';
            const stdout = err.stdout?.toString() || '';
            throw new Error(`drizzle-kit push failed:\n${stderr || stdout || err.message}`);
        }
        // Check for error patterns in output even if exit code is 0
        if (output && (output.includes('error:') || output.includes('Error:'))) {
            throw new Error(`drizzle-kit push failed:\n${output}`);
        }
        if (verbose && output) {
            console.log(output);
        }
        if (verbose) {
            console.log('✅ Database migration completed successfully');
        }
    }
    finally {
        if (fs.existsSync(tempConfigPath)) {
            fs.unlinkSync(tempConfigPath);
        }
    }
}
