// coze-coding-dev-sdk@0.5.5/dist/search/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchClient = void 0;
const client_1 = require("../core/client");
const exceptions_1 = require("../core/exceptions");
const models_1 = require("./models");
class SearchClient extends client_1.BaseClient {
    constructor(config, customHeaders, verbose = false) {
        super(config, customHeaders, verbose);
    }
    async search(request) {
        const apiRequest = (0, models_1.convertToApiFormat)(request);
        const response = await this.request('POST', `${this.config.baseUrl}/api/search_api/web_search`, apiRequest);
        const responseMetadata = response.ResponseMetadata || {};
        if (responseMetadata.Error) {
            throw new exceptions_1.APIError(`Search failed: ${responseMetadata.Error}`, 500, responseMetadata.Error);
        }
        const result = response.Result || {};
        const webItems = [];
        if (result.WebResults) {
            for (const item of result.WebResults) {
                webItems.push((0, models_1.convertFromApiFormat)(item));
            }
        }
        const imageItems = [];
        if (result.ImageResults) {
            for (const item of result.ImageResults) {
                imageItems.push((0, models_1.convertFromApiFormat)(item));
            }
        }
        let summary;
        if (result.Choices && result.Choices.length > 0) {
            summary = result.Choices[0].Message?.Content;
        }
        return {
            web_items: webItems,
            image_items: imageItems,
            summary,
        };
    }
    async webSearch(query, count = 10, needSummary = true) {
        return this.search({
            query,
            search_type: 'web',
            count,
            need_summary: needSummary,
        });
    }
    async webSearchWithSummary(query, count = 10) {
        return this.search({
            query,
            search_type: 'web_summary',
            count,
            need_summary: true,
        });
    }
    async imageSearch(query, count = 10) {
        return this.search({
            query,
            search_type: 'image',
            count,
            need_summary: false,
        });
    }
    async advancedSearch(query, options = {}) {
        const filter = {
            need_content: options.needContent,
            need_url: options.needUrl,
            sites: options.sites,
            block_hosts: options.blockHosts,
        };
        return this.search({
            query,
            search_type: options.searchType || 'web',
            count: options.count || 10,
            filter,
            need_summary: options.needSummary !== undefined ? options.needSummary : true,
            time_range: options.timeRange,
        });
    }
}
exports.SearchClient = SearchClient;
