// coze-coding-dev-sdk@0.5.5/dist/search/models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertToApiFormat = convertToApiFormat;
exports.convertFromApiFormat = convertFromApiFormat;
function convertToApiFormat(data) {
    const result = {};
    if (data.query) {
        result.Query = data.query;
    }
    if (data.search_type) {
        result.SearchType = data.search_type;
    }
    if (data.count !== undefined) {
        result.Count = data.count;
    }
    if (data.need_summary !== undefined) {
        result.NeedSummary = data.need_summary;
    }
    if (data.time_range) {
        result.TimeRange = data.time_range;
    }
    if (data.filter) {
        result.Filter = {
            NeedContent: data.filter.need_content || false,
            NeedUrl: data.filter.need_url || false,
            Sites: data.filter.sites,
            BlockHosts: data.filter.block_hosts,
        };
    }
    return result;
}
function convertFromApiFormat(item) {
    const result = {};
    if (item.Id !== undefined)
        result.id = item.Id;
    if (item.SortId !== undefined)
        result.sort_id = item.SortId;
    if (item.Title !== undefined)
        result.title = item.Title;
    if (item.SiteName !== undefined)
        result.site_name = item.SiteName;
    if (item.Url !== undefined)
        result.url = item.Url;
    if (item.Snippet !== undefined)
        result.snippet = item.Snippet;
    if (item.Summary !== undefined)
        result.summary = item.Summary;
    if (item.Content !== undefined)
        result.content = item.Content;
    if (item.PublishTime !== undefined)
        result.publish_time = item.PublishTime;
    if (item.LogoUrl !== undefined)
        result.logo_url = item.LogoUrl;
    if (item.RankScore !== undefined)
        result.rank_score = item.RankScore;
    if (item.AuthInfoDes !== undefined)
        result.auth_info_des = item.AuthInfoDes;
    if (item.AuthInfoLevel !== undefined)
        result.auth_info_level = item.AuthInfoLevel;
    if (item.Image) {
        result.image = {
            url: item.Image.Url,
            width: item.Image.Width,
            height: item.Image.Height,
            shape: item.Image.Shape,
        };
    }
    return result;
}
